ToolChain/Assembler/Parser/Parser.py

```diff
@@ -20,586 +19,411 @@
+        the binary values for each register and immediate value. Labels are converted to bytestring as they are.
+        :param ins: str, the actual instruction itself
+        :param state: str, State of the instruction based on evaluated form (ex: InsReg = STATE1)
+        :param operandList: str list, List of operands to be converted to binary code and assembled
+        :return: bytestring, The newly built binary code for the instruction and operands
+        """
 
-            if type(instructionPartOriginal) is str:
-                # This is a memory reference
-                memoryRefSymbol = instructionPartOriginal
-                instructionPartToBeUsed = 0xDEADBEEF
+        instruction = b''
+
+        if state == "STATE0":
+            # Instruction is already complete, there is only one operand (the instruction itself, Ins)
+            pass
+
+        elif state == "STATE1":
+            # Form = Instruction - Register
+            register = self.translateRegisterName(operandList[0][1:])
+            instruction += bytes(0b0000) + bytes((register,))
+
+        elif state == "STATE2":
+            # Form = Instruction - Register - Register
+            register = self.translateRegisterName(operandList[0][1:])
+            register2 = self.translateRegisterName(operandList[1][1:])
+            register = (register << 4) + register2
+            instruction += bytes((register,))
+
+        elif state == "STATE3":
+            # Form = Instruction - Immediate
+            if operandList[0][0] == "#":
+                immediate = self.translateTextImmediate(operandList[0][1:])
+                instruction += immediate.to_bytes(4, byteorder='big')
             else:
-                instructionPartToBeUsed = instructionPartOriginal
-
-            instructionPartBinaryValue = instructionPartBitMask & instructionPartToBeUsed
-            binaryInstruction |= (instructionPartBinaryValue << shiftValue)
-
-        # At this point, logic is we have a fully built binary instruction.
-        # We now need to transform this into a byte string
-        binaryInstructionByteString = b""
+                if operandList[0].count(MEMORY_REFERENCE_INDICATORS) == 0:
+                    instruction += b':' + operandList[0].encode("utf-8") + b':'
+                else:
+                    raise ValueError("Syntax error, memory reference may not contain any colons in its name - \":\"")
+
+        elif state == "STATE4":
+            # Form = Instruction - Immediate - Register
+            register = self.translateRegisterName(operandList[1][1:])
+            if operandList[0][0] == "#":
+                immediate = self.translateTextImmediate(operandList[0][1:])
+                instruction += immediate.to_bytes(4, byteorder='big') + bytes(0b0000) + bytes((register,))
+            else:
+                if operandList[0].count(MEMORY_REFERENCE_INDICATORS) == 0:
```

Comment:
This chunk of code is reused at a high number of place in this larger block of code. I would do a function with this to simplify the code.